cpu/esp32/periph/pm.c

```diff
@@ -41,65 +44,127 @@ void pm_set_lowest(void)
-    rtc_sleep_init(config);
-    rtc_sleep_start(RTC_TIMER_TRIG_EN, 0);
+    DEBUG ("%s enter to power mode %d @%u\n", __func__, mode, system_get_time());
 
-    /* Restore CPU frequency */
-    rtc_clk_cpu_freq_set(cpu_freq);
+    /* Labels for RTC slow memory that are defined in the linker script */
+    extern int _rtc_bss_rtc_start;
+    extern int _rtc_bss_rtc_end;
 
-    /* resume UARTs */
-    for (int i = 0; i < 3; ++i) {
-        REG_CLR_BIT(UART_FLOW_CONF_REG(i), UART_FORCE_XOFF);
-        REG_SET_BIT(UART_FLOW_CONF_REG(i), UART_FORCE_XON);
-        REG_CLR_BIT(UART_FLOW_CONF_REG(i), UART_FORCE_XON);
+    /*
+     * Activate the Power Domain for slow RTC memory when the .rtc.bss
+     * section is used to retain uninitialized data. The Power Domain for
+     * slow RTC memory is automatically activated when the .rtc.data section
+     * is used to retain initialized data.
+     */
+    if (&_rtc_bss_rtc_end > &_rtc_bss_rtc_start) {
+        esp_sleep_pd_config(ESP_PD_DOMAIN_RTC_SLOW_MEM, ESP_PD_OPTION_ON);
     }
-}
 
-extern void esp_restart_noos(void) __attribute__ ((noreturn));
+    /*
+     * Activate the power domain for RTC peripherals either when
+     * ESP_PM_GPIO_HOLD is defined or when light sleep mode is activated.
+     * As long as the RTC peripherals are active, the pad state of RTC GPIOs
+     * is held in deep sleep and the pad state of all GPIOs is held in light
+     * sleep.
+     */
+#ifdef ESP_PM_GPIO_HOLD
+    esp_sleep_pd_config(ESP_PD_DOMAIN_RTC_PERIPH, ESP_PD_OPTION_ON);
+#else
+    if (mode == ESP_PM_LIGHT_SLEEP) {
+        esp_sleep_pd_config(ESP_PD_DOMAIN_RTC_PERIPH, ESP_PD_OPTION_ON);
+    }
+#endif
 
-void pm_reboot(void)
-{
-    DEBUG ("%s\n", __func__);
+    /* Prepare the RTC timer if an RTC alarm is set to wake up. */
+    struct tm tm_alarm;
+    struct tm tm_system;
 
-    /* suspend and flush UARTs */
-    for (int i = 0; i < 3; ++i) {
-        REG_SET_BIT(UART_FLOW_CONF_REG(i), UART_FORCE_XOFF);
-        uart_tx_wait_idle(i);
+    rtc_get_time(&tm_system);
+    rtc_get_alarm(&tm_alarm);
+
+    time_t t_system = mktime(&tm_system);
+    time_t t_alarm = mktime(&tm_alarm);
+    int _alarm_set = 0;
+
+    if (t_alarm > t_system) {
+        _alarm_set = 1;
+        esp_sleep_enable_timer_wakeup((uint64_t)(t_alarm - t_system) * US_PER_SEC);
     }
 
-    software_reset();
+#ifdef ESP_PM_WUP_PINS
+    /*
+     * Prepare the wake-up pins if a single pin or a comma-separated list of
+     * pins is defined for wake-up.
+     */
+    static const gpio_t wup_pins[] = { ESP_PM_WUP_PINS };
+
+    uint64_t wup_pin_mask = 0;
+    for (unsigned i = 0; i < ARRAY_SIZE(wup_pins); i++) {
+        wup_pin_mask |= 1ULL << wup_pins[i];
+    }
+#ifdef ESP_PM_WUP_LEVEL
+   esp_sleep_enable_ext1_wakeup(wup_pin_mask, ESP_PM_WUP_LEVEL);
+#else /* ESP_PM_WUP_LEVEL */
+   esp_sleep_enable_ext1_wakeup(wup_pin_mask, ESP_EXT1_WAKEUP_ANY_HIGH);
+#endif /* ESP_PM_WUP_LEVEL */
+#endif /* ESP_PM_WUP_PINS */
```

Comment:
Couldn't this be done in `gpio_init_int()`?
The you could also drop the need of configuring the wake-up GPIOs at compile time.

Reply:
> Couldn't this be done in gpio_init_int()?

You mean for GPIOs (if they are RTC GPIOs) that are configured as `GPIO_IN` could also be used as wake-up source? That's not a good idea. It must be clear to the user that GPIOs in light/deep sleep mode are handled differently than usual. GPIOs cannot be controlled individually in sleep modes. A wake-up occurs when either all GPIOs become LOW (`ESP_EXT1_WAKEUP_ALL_LOW`) or any GPIO becomes HIGH (`ESP_EXT1_WAKEUP_ANY_HIGH`).

For example, if you want to wake up by a LOW active interrupt line, `ESP_EXT1_WAKEUP_ALL_LOW` would have to be selected as wake-up level. Then you would have to make sure that all other GPIOs configured as `GPIO_IN` are already LOW. This is hardly controllable. Using an explicit setting of the only one GPIO that is used as wake-up source is much easier in that case.

Unfortunately, the wake-up functions in ESP32 are not so sophisticated. It is something like all or nothing. Multiple GPIOs can only be used as a wake-up source if, for example, all connected interrupt lines were HIGH active, which is normally not the case.

So the actual use case will be to use one GPIO for an external interrupt line to wake up the ESP32 ([one example](https://github.com/RIOT-OS/RIOT/issues/13365#issuecomment-593862248)). This can then be LOW or HIGH active.

Reply:
@benpicco Unfortunately I can neither find the reason nor a solution for the problem that the GPIO interrupt is triggered in addition to the RTC alarm when leaving the sleep mode the first time.

However, in a newer version of the SDK I have found a way to use GPIOs configured as input with interrupts directly as wakeup source. After merging the changes from new SDK version in the SDK version used by RIOT, I can use GPIOs directly as interrupt source :smile:

Unfortunately only level interrupts and no edge interrupts are supported. What a bummer.  I'm stumped. Ok, we could try to save the interrupt type of all GPIOs and map the edge interrupts to level interrupts before entering sleep mode and change them back when returning from the sleep mode. However, a unique mapping is difficult. Normally you would map `GPIO_RISING` to `GPIO_HIGH` and `GPIO_FALLING` to `GPIO_LOW` if it's only about waking up. Unfortunately, this does not work with our button in the `tests/periph_pm`, because `GPIO_RISING` is used, but the ESP32 button is LOW active, so it generates an interrupt when it is released and the mapping above would lead to `GPIO_HIGH` as wakeup level.